osp/core/cuds.py

```diff
@@ -1,106 +1,146 @@
+                 iri: URIRef = None,
+                 uid: UUID = None):
+        """Initialize a CUDS object."""
+        # Set identifier. This is a "user-facing" method, so strict types
+        # checks are performed.
+        if oclass and attributes is None:
+            raise TypeError("__init__() missing 1 optional argument:"
+                            "`attributes`. This attribute is required when"
+                            "an ontology class is specified.")
+        if identifier is not None and any(x is not None for x in (iri, uid)):
+            raise Exception("The keyword argument `identifier` cannot be used"
+                            "at the same time as the keyword arguments `iri`"
+                            "or `uuid`.")
+        elif all(x is not None for x in (iri, uid)):
+            raise Exception("Tried to initialize a CUDS object specifying, "
+                            "both its IRI and its UID. A CUDS object is "
+                            "constrained to have just one identifier.")
+        elif identifier is not None \
+                and type(identifier) not in (UUID, URIRef):
+            raise ValueError('Provide either a UUID or a URIRef object.')
+        elif uid is not None and type(uid) is not UUID:
+            raise ValueError('Provide a UUID object.')
+        elif iri is not None and type(iri) is not URIRef:
+            raise ValueError('Provide a URIRef object.')
+        else:
+            self._identifier = identifier or uid or iri or uuid4()
+
+        # Create CUDS triples in internal temporary graph.
+        self._graph = Graph()
+        if attributes:
+            for k, v in attributes.items():
+                self._graph.add((
+                    self.iri, k.iri, Literal(k.convert_to_datatype(v),
+                                             datatype=k.datatype)
+                ))
         if oclass:
             self._graph.add((
-                self.iri, rdflib.RDF.type, oclass.iri
+                self.iri, RDF.type, oclass.iri
             ))
+
+        self._session = session or Cuds._session
+        # Copy temporary graph to the session graph and discard it.
         self.session._store(self)
 
     @property
-    def uid(self) -> uuid.UUID:
-        """Get he uid of the cuds object."""
-        return self.__uid
+    def uid(self) -> Union[URIRef, UUID]:
+        """Get the identifier of the CUDS object.
+
+        Does just the same as the `identifier` property, it just exists for
+        backwards compatibility purposes (with the wrappers).
+        """
+        return self.identifier
 
     @property
-    def iri(self):
+    def iri(self) -> URIRef:
         """Get the IRI of the CUDS object."""
-        return iri_from_uid(self.uid)
+        return self.identifier if type(self.identifier) is URIRef else \
+            URIRef(CUDS_NAMESPACE_IRI + str(self.identifier))
+
+    @property
+    def identifier(self) -> Union[URIRef, UUID]:
```

Comment:
Isn't better to have `id` that is shorter than `identifier`?

Reply:
In general, I am not a friend of very short abbreviations in code, specially having autocompletion. In this case, it is still easy to recognize the meaning of abbreviation though. If you think it is really bad to have the longer version then I can of course change it.